packages/services/src/providers/spot-live/rpc-spot.provider.ts

```diff
@@ -0,0 +1,775 @@
+      if (existing) {
+        existing.push(normalizedTransfer);
+      } else {
+        byTx.set(transfer.hash, [normalizedTransfer]);
+      }
+    }
+
+    return byTx;
+  }
+
+  /**
+   * Detect swap pattern from transaction transfers
+   * A swap has both outbound and inbound transfers for the same wallet
+   * @param transfers Transfers in a single transaction
+   * @param walletAddress Wallet address being monitored
+   * @returns Detected swap or null if not a swap pattern
+   */
+  private detectSwapPattern(
+    transfers: Transfer[],
+    walletAddress: string,
+  ): DetectedSwap | null {
+    const wallet = walletAddress.toLowerCase();
+
+    // Separate outbound and inbound transfers
+    const outbound = transfers.filter((t) => t.from === wallet);
+    const inbound = transfers.filter((t) => t.to === wallet);
+
+    // Swap requires at least one outbound and one inbound
+    if (outbound.length === 0 || inbound.length === 0) {
+      return null;
+    }
+
+    // Use first outbound and first inbound (handles most cases)
+    // Note: Multi-hop swaps may have multiple transfers, taking first of each is reasonable heuristic
+    const out = outbound[0];
+    const inc = inbound[0];
+
+    if (!out || !inc) {
+      return null;
+    }
+
+    const fromToken = this.getTokenAddress(out);
+    const toToken = this.getTokenAddress(inc);
+
+    return {
+      txHash: out.hash,
+      blockNumber: parseInt(out.blockNum, 16),
+      timestamp: new Date(out.metadata.blockTimestamp),
+      fromToken,
+      toToken,
+      fromAmount: out.value,
+      toAmount: inc.value,
+      protocol: "Unknown", // Will be set by protocol filter or left as Unknown
+    };
+  }
+
+  /**
+   * Two-layer protocol filtering (chain-specific)
+   * Layer 1: Router address check (fast rejection)
+   * Layer 2: Swap event signature verification (validation)
+   * @param txHash Transaction hash to check
+   * @param chain Chain where transaction occurred
+   * @param chainFilters Protocol filters for this chain
+   * @returns Filter result with allowed flag and protocol name if matched
+   */
+  private async checkProtocolFilter(
+    txHash: string,
+    chain: SpecificChain,
+    chainFilters: ProtocolFilter[],
+  ): Promise<{ allowed: boolean; protocol?: string }> {
+    try {
+      // Fetch transaction and receipt in parallel
+      const [tx, receipt] = await Promise.all([
+        this.rpcProvider.getTransaction(txHash, chain),
+        this.rpcProvider.getTransactionReceipt(txHash, chain),
+      ]);
+
+      if (!tx || !receipt) {
+        this.logger.debug(
+          {
+            txHash,
+            chain,
+          },
+          `[RpcSpotProvider] Could not fetch tx/receipt`,
+        );
+        return { allowed: false };
+      }
+
+      // Check each protocol filter for this chain
+      for (const filter of chainFilters) {
+        // Layer 1: Router address check (fast)
+        if (!tx.to) {
```

Comment:
I didn't super confirm this one so take it with a grain of salt: 
Protocol filtering checks tx.to (rpc-spot.provider.ts:583), which doesn't exist in Solana transactions
Solana uses a different transaction model (programs, SPL tokens, no to address)

Reply:
Was considering building in support for solana if time permitted for spot live but ended up choosing to block it in https://github.com/recallnet/js-recall/pull/1596 - see commit [558247b](https://github.com/recallnet/js-recall/pull/1596/commits/558247b2c23b71fc848af682b73252448428f190)